Warning and build failure in archive_write_set_format_ar

- Version of libarchive (master at 6c739e0a968628063f94ecb19db3db6a5854dcde )
- built from source
- compiler: gcc 8.1.0
- system: linux, archlinux.

I am getting the following warnings, these turned into errors when using libarchive (and compiling it) as an external library in other project.

I guess due to libarchive turning these warnings into errors in [CMakeLists.txt](https://github.com/libarchive/libarchive/blob/master/CMakeLists.txt)

```
[220/666] Building C object libarchive/CMakeFiles/archive.dir/archive_write_set_format_ar.c.o
/path/Software/Slicer/libarchive/libarchive/libarchive/archive_write_set_format_ar.c: In function ‘archive_write_ar_header’:
/path/Software/Slicer/libarchive/libarchive/libarchive/archive_write_set_format_ar.c:183:2: warning: ‘strncpy’ output truncated before terminating nul copying 2 bytes from a string of the same length [-Wstringop-truncation]
  strncpy(&buff[AR_fmag_offset], "`\n", 2);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/libarchive/archive_write_set_format_ar.c:192:3: warning: ‘strncpy’ output truncated before terminating nul copying 9 bytes from a string of the same length [-Wstringop-truncation]
   strncpy(buff + AR_name_offset, "__.SYMDEF", 9);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/libarchive/archive_write_set_format_ar.c:292:4: warning: ‘strncpy’ output truncated before terminating nul copying 3 bytes from a string of the same length [-Wstringop-truncation]
    strncpy(buff + AR_name_offset, "#1/", 3);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/libarchive/archive_write_set_format_ar.c:288:4: warning: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
    strncpy(&buff[AR_name_offset], filename, strlen(filename));
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/libarchive/archive_write_set_format_ar.c:287:7: note: length computed here
   if (strlen(filename) <= 16 && strchr(filename, ' ') == NULL) {
       ^~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/libarchive/archive_write_set_format_ar.c:251:4: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
    strncpy(se, filename, strlen(filename));
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/libarchive/archive_write_set_format_ar.c:228:4: warning: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
    strncpy(&buff[AR_name_offset],
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        filename, strlen(filename));
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/libarchive/archive_write_set_format_ar.c:227:7: note: length computed here
   if (strlen(filename) <= 15) {
       ^~~~~~~~~~~~~~~~
[267/666] Building C object libarchive/test/CMakeFiles/libarchive_test.dir/__/__/test_utils/test_main.c.o
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c: In function ‘main’:
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:26: warning: ‘%s’ directive writing up to 255 bytes into a region of size between 250 and 254 [-Wformat-overflow=]
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
                          ^~
       tmpdir_timestamp, i);
       ~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:19: note: directive argument in the range [0, 999]
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
                   ^~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:3: note: ‘sprintf’ output 7 or more bytes (assuming 266) into a destination of size 256
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       tmpdir_timestamp, i);
       ~~~~~~~~~~~~~~~~~~~~
[551/666] Building C object cat/test/CMakeFiles/bsdcat_test.dir/__/__/test_utils/test_main.c.o
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c: In function ‘main’:
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:26: warning: ‘%s’ directive writing up to 255 bytes into a region of size between 250 and 254 [-Wformat-overflow=]
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
                          ^~
       tmpdir_timestamp, i);
       ~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:19: note: directive argument in the range [0, 999]
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
                   ^~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:3: note: ‘sprintf’ output 7 or more bytes (assuming 266) into a destination of size 256
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       tmpdir_timestamp, i);
       ~~~~~~~~~~~~~~~~~~~~
[588/666] Building C object tar/test/CMakeFiles/bsdtar_test.dir/__/__/test_utils/test_main.c.o
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c: In function ‘main’:
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:26: warning: ‘%s’ directive writing up to 255 bytes into a region of size between 250 and 254 [-Wformat-overflow=]
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
                          ^~
       tmpdir_timestamp, i);
       ~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:19: note: directive argument in the range [0, 999]
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
                   ^~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:3: note: ‘sprintf’ output 7 or more bytes (assuming 266) into a destination of size 256
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       tmpdir_timestamp, i);
       ~~~~~~~~~~~~~~~~~~~~
[653/666] Building C object cpio/test/CMakeFiles/bsdcpio_test.dir/test_option_t.c.o
/path/Software/Slicer/libarchive/libarchive/cpio/test/test_option_t.c: In function ‘test_option_t’:
/path/Software/Slicer/libarchive/libarchive/cpio/test/test_option_t.c:95:38: warning: ‘ file’ directive output may be truncated writing 5 bytes into a region of size between 0 and 19 [-Wformat-truncation=]
  snprintf(date, sizeof(date)-1, "%12s file", date2);
                                      ^~~~~
/path/Software/Slicer/libarchive/libarchive/cpio/test/test_option_t.c:95:2: note: ‘snprintf’ output between 18 and 37 bytes into a destination of size 31
  snprintf(date, sizeof(date)-1, "%12s file", date2);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[660/666] Building C object cpio/test/CMakeFiles/bsdcpio_test.dir/__/__/test_utils/test_main.c.o
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c: In function ‘main’:
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:26: warning: ‘%s’ directive writing up to 255 bytes into a region of size between 250 and 254 [-Wformat-overflow=]
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
                          ^~
       tmpdir_timestamp, i);
       ~~~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:19: note: directive argument in the range [0, 999]
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
                   ^~~~~~~~~~~~~~~
/path/Software/Slicer/libarchive/libarchive/test_utils/test_main.c:3737:3: note: ‘sprintf’ output 7 or more bytes (assuming 266) into a destination of size 256
   sprintf(tmpdir, "%s/%s.%s-%03d", tmp, progname,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       tmpdir_timestamp, i);

```

--- a/libarchive/archive_write_set_format_ar.c
+++ b/libarchive/archive_write_set_format_ar.c
@@ -180,7 +180,7 @@ archive_write_ar_header(struct archive_write *a, struct archive_entry *entry)
 	}
 
 	memset(buff, ' ', 60);
-	strncpy(&buff[AR_fmag_offset], "`\n", 2);
+	memcpy(&buff[AR_fmag_offset], "`\n", 2);
 
 	if (strcmp(pathname, "/") == 0 ) {
 		/* Entry is archive symbol table in GNU format */
@@ -189,7 +189,7 @@ archive_write_ar_header(struct archive_write *a, struct archive_entry *entry)
 	}
 	if (strcmp(pathname, "__.SYMDEF") == 0) {
 		/* Entry is archive symbol table in BSD format */
-		strncpy(buff + AR_name_offset, "__.SYMDEF", 9);
+		memcpy(buff + AR_name_offset, "__.SYMDEF", 9);
 		goto stat;
 	}
 	if (strcmp(pathname, "//") == 0) {
@@ -225,7 +225,7 @@ archive_write_ar_header(struct archive_write *a, struct archive_entry *entry)
 		 * actually 15 bytes.
 		 */
 		if (strlen(filename) <= 15) {
-			strncpy(&buff[AR_name_offset], 
+			memcpy(&buff[AR_name_offset],
 			    filename, strlen(filename));
 			buff[AR_name_offset + strlen(filename)] = '/';
 		} else {
@@ -248,7 +248,7 @@ archive_write_ar_header(struct archive_write *a, struct archive_entry *entry)
 				return (ARCHIVE_FATAL);
 			}
 
-			strncpy(se, filename, strlen(filename));
+			memcpy(se, filename, strlen(filename));
 			strcpy(se + strlen(filename), "/\n");
 
 			ss = strstr(ar->strtab, se);
@@ -285,11 +285,11 @@ archive_write_ar_header(struct archive_write *a, struct archive_entry *entry)
 		 * archive header.
 		 */
 		if (strlen(filename) <= 16 && strchr(filename, ' ') == NULL) {
-			strncpy(&buff[AR_name_offset], filename, strlen(filename));
+			memcpy(&buff[AR_name_offset], filename, strlen(filename));
 			buff[AR_name_offset + strlen(filename)] = ' ';
 		}
 		else {
-			strncpy(buff + AR_name_offset, "#1/", 3);
+			memcpy(buff + AR_name_offset, "#1/", 3);
 			if (format_decimal(strlen(filename),
 			    buff + AR_name_offset + 3,
 			    AR_name_size - 3)) {

